ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/api/UsesEditorConfigProperties.kt

```diff
@@ -0,0 +1,54 @@
+package com.pinterest.ktlint.core.api
+
+import com.pinterest.ktlint.core.EditorConfig
+
+/**
+ * Indicates [com.pinterest.ktlint.core.Rule] uses properties loaded from `.editorconfig` file.
+ *
+ * This properties could be:
+ * - universal `.editorconfig` properties defined
+ * [here](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#current-universal-properties)
+ * - universal IntelliJ IDEA properties defined
+ * [here](https://github.com/JetBrains/intellij-community/blob/master/platform/lang-api/src/com/intellij/psi/codeStyle/CommonCodeStyleSettings.java)
+ * - Kotlin specific properties defined
+ * [here](https://github.com/JetBrains/kotlin/blob/master/idea/formatter/src/org/jetbrains/kotlin/idea/core/formatter/KotlinCodeStyleSettings.java)
+ *
+ * In the best case rule should only use one property.
+ *
+ * See [com.pinterest.ktlint.core.KtLint.generateKotlinEditorconfigSection] documentation how to generate
+ * `.editorconfig` based on [com.pinterest.ktlint.core.Rule]s with this interface implementations.
+ */
+@FeatureInAlphaState
+public interface UsesEditorConfigProperties {
+
+    /**
+     * Provide a list of code style editorconfig properties, that rule validates.
+     *
+     * @return list of properties, for example:
+     * ```
+     * listOf("indent_size", "ij_kotlin_else_on_new_line")
+     * ```
+     */
+    public val editorconfigProperties: List<String>
```

Comment:
Did you think of introducing something like
```
data class EditorConfigProperty<T>(
  val key: String,
  val value: T?,
  val defaultValue: T
)
```
then we'd only need to return a list of those in this interface. Ideally `EditorConfig` would also contain those

Reply:
No, I haven't. Will think if it is applicable here :thinking:

Reply:
@romtsn please check 9025377 commit.

I haven't introduce `<T>` generic as editorconfig loading becomes too complex (imho) and `T` is only used inside rule. Rule could convert String internally into any representation.

Reply:
@romtsn and here is another new approach that uses ec4j `PropertyType`: fc1ab0f